include/ResultBase.hpp

```diff
@@ -0,0 +1,20 @@
+//
+// Created by Tianyi Wang on 5/15/23.
+//
+
+#ifndef DD_EVAL_RESULTBASE_HPP
+#define DD_EVAL_RESULTBASE_HPP
+
+#include <nlohmann/json.hpp>
+
+using json = nlohmann::json;
+
+class ResultBase {
+public:
+  virtual ~ResultBase() = default;
+
+  virtual json        toJson() const   = 0;
+  virtual std::string toString() const = 0;
```

Comment:
```suggestion
  virtual std::string toString() const { return toJson().dump(2); }
```
No need to make this function abstract.